samples/bluetooth/peripheral_uart/src/uart_async_adapter.c

```diff
@@ -0,0 +1,551 @@
+		uart_irq_tx_disable(data->target);
+
+		k_spinlock_key_t key = k_spin_lock(&(data->lock));
+		struct uart_event event = {UART_TX_DONE};
+
+		if (!data->tx.buf) {
+			k_spin_unlock(&(data->lock), key);
+			LOG_ERR("TX buffer NULL on tx complete");
+		} else {
+			__ASSERT_NO_MSG(data->tx.curr_buf);
+			__ASSERT_NO_MSG(data->tx.curr_buf >= data->tx.buf);
+			/* Transfer really finished */
+			event.data.tx.buf = data->tx.buf;
+			event.data.tx.len = data->tx.curr_buf - data->tx.buf;
+			/* Clear any pending tx */
+			data->tx.buf = NULL;
+			data->tx.curr_buf = NULL;
+			data->tx.size_left = 0;
+			k_spin_unlock(&(data->lock), key);
+			LOG_DBG("Notification: UART_TX_DONE (0x%x, size: %d)",
+				(unsigned int)event.data.tx.buf, event.data.tx.len);
+			user_callback(dev, &event);
+		}
+	}
+	LOG_DBG("%s: Exit", __func__);
+}
+
+static inline void on_rx_ready(const struct device *dev, struct uart_async_adapter_data *data)
+{
+	int ret;
+	bool notify_now = false;
+
+	LOG_DBG("%s: Enter (%s)", __func__, dev->name);
+	if (data->rx.timeout != SYS_FOREVER_MS) {
+		k_timer_start(&data->rx.timeout_timer, K_MSEC(data->rx.timeout), K_NO_WAIT);
+	}
+	do {
+		k_spinlock_key_t key = k_spin_lock(&(data->lock));
+
+		if (!data->rx.size_left) {
+			notify_now = false;
+			k_spin_unlock(&(data->lock), key);
+			notify_rx_buffer(dev);
+			switch_rx_buffer(dev, true);
+			key = k_spin_lock(&(data->lock));
+		}
+		if (!data->rx.size_left) {
+			/* Data received without buffer - dropping */
+			uint8_t dummy;
+			size_t cnt = 0;
+
+			do {
+				ret = uart_fifo_read(data->target, &dummy, 1);
+				if (ret < 0) {
+					LOG_ERR("Unexpected error on FIFO dropping: %d", ret);
+					ret = 0;
+				}
+				cnt += ret;
+			} while (ret);
+			LOG_ERR("Data received without buffer prepared, dropped %d bytes", cnt);
+		} else {
+			ret = uart_fifo_read(data->target, data->rx.curr_buf, data->rx.size_left);
+			LOG_DBG("Received %d characters", ret);
+			if (ret < 0) {
+				LOG_ERR("Unexpected error on FIFO read: %d", ret);
+				ret = 0;
+			}
+			__ASSERT_NO_MSG(data->rx.size_left >= ret);
+			data->rx.curr_buf += ret;
+			data->rx.size_left -= ret;
+			if (data->rx.timeout == 0) {
+				notify_now = true;
+			}
+		}
+		k_spin_unlock(&(data->lock), key);
+	} while (ret);
+	if (notify_now) {
+		notify_rx_buffer(dev);
+	}
+	LOG_DBG("%s: Exit", __func__);
+}
+
+static inline void on_error(const struct device *dev,
+			    struct uart_async_adapter_data *data,
+			    int rx_err)
+{
+	struct uart_event event = {
+		.type = UART_RX_STOPPED,
+		.data.rx_stop.reason = rx_err
+	};
+
+	LOG_DBG("%s: Enter(%s)", __func__, dev->name);
+	rx_disable(data->target);
+	user_callback(dev, &event);
+	LOG_DBG("%s: Exit", __func__);
+}
+
+static void uart_irq_handler(const struct device *target_dev, void *context)
+{
+	const struct device *dev = context;
+	struct uart_async_adapter_data *data = access_dev_data(dev);
+
+	__ASSERT(target_dev == data->target, "FUBOR in uart_async_adapter initialization");
```

Comment:
Fatal Unhandled Behavior Ordering Reset - FUBOR
Standard thing... probably not used frequently these days though...

Reply:
Fucked Up Beyond Of Repair. But right - it is like an "ass" left in the code.
Changed to:
"IRQ handler called with a context that seems uninitialized."

Reply:
Technically it should never hit this assert. If it would - it means something really deeply broken in the code. If not the driver or adapter - something like memory leak. I expect that if this happens whole system would be unstable anyway.